overlord/snapstate/handlers.go

```diff
@@ -1989,26 +1989,26 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 	// find if the snap is already installed before we modify snapst below
 	isInstalled := snapst.IsInstalled()
 
-	cand := snapsup.SideInfo
-	m.backend.Candidate(cand)
+	cand := NewRevisionSideInfo(snapsup.SideInfo, nil)
+	m.backend.Candidate(cand.SnapSideInfo)
 
-	oldCandidateIndex := snapst.LastIndex(cand.Revision)
+	oldCandidateIndex := snapst.LastIndex(cand.SnapSideInfo.Revision)
 
 	var oldRevsBeforeCand []snap.Revision
 	if oldCandidateIndex < 0 {
-		snapst.Sequence = append(snapst.Sequence, cand)
+		snapst.Sequence.Revisions = append(snapst.Sequence.Revisions, cand)
 	} else if !snapsup.Revert {
 		// save the revs before the candidate, so undoLink can account for discarded revs when putting it back
-		for _, si := range snapst.Sequence[:oldCandidateIndex] {
-			oldRevsBeforeCand = append(oldRevsBeforeCand, si.Revision)
+		for _, si := range snapst.Sequence.Revisions[:oldCandidateIndex] {
+			oldRevsBeforeCand = append(oldRevsBeforeCand, si.SnapSideInfo.Revision)
 		}
 		// remove the old candidate from the sequence, add it at the end
-		copy(snapst.Sequence[oldCandidateIndex:len(snapst.Sequence)-1], snapst.Sequence[oldCandidateIndex+1:])
-		snapst.Sequence[len(snapst.Sequence)-1] = cand
+		copy(snapst.Sequence.Revisions[oldCandidateIndex:len(snapst.Sequence.Revisions)-1], snapst.Sequence.Revisions[oldCandidateIndex+1:])
+		snapst.Sequence.Revisions[len(snapst.Sequence.Revisions)-1] = cand
 	}
```

Comment:
it would be good to start thinking about turning some of this logic into methods of SnapSequence or SnapState taking needed parameters, same of other bits like this in this file. Anything digging into Sequence to do copy or append is a bit suspicious after the change

Reply:
Right, I've added a TODO in the `SnapSequence` definition about this. This can be followed up in next components PRs.